File: leetcode/find_anagrams.py
#!/usr/bin/env python3

from collections import defaultdict
import sys
import unittest

# number: 438
# similar: 567
# title: Find All Anagrams in a String
# url: https://leetcode.com/problems/find-all-anagrams-in-a-string/
# section: meta
# difficulty: medium
# tags: hash table, string, sliding window, meta, grind 75

# constraints
# 1 <= s.length, p.length <= 3 * 10^4
# s and p consist of lowercase English letters.

# solution: LeetCode solution, sliding window + defaultdict
# complexity
# run-time: O(n)
# space: O(1) (only 26 letters per dict)
def find_anagrams(s: str, p: str) -> list[int]:
    if len(p) > len(s):
        return []
    
    pmap = defaultdict(int)
    
    for c in p:
        pmap[c] += 1
        
    # sliding window counter
    winmap = defaultdict(int)
    left = 0
    ans = []

    for right in range(len(s)):
        winmap[s[right]] += 1
        
        # window is smaller than p
        if right-left+1 < len(p):
            continue
            
        # check if the current window is a permutation of p
        if pmap == winmap:
            ans.append(left)
        
        # remove the leftmost character from the window
        if winmap[s[left]] > 1:
            winmap[s[left]] -= 1
        else:
            del winmap[s[left]]
            
        # slide the window
        left += 1
            
    return ans

class TestFindAnagram(unittest.TestCase):
    def test_empty(self):
        s = ""
        p = ""
        expected = []
        self.assertEqual(find_anagrams(s, p), expected)

    def test_1(self):
        s = "cbaebabacd"
        p = "abc"
        expected = [0, 6]
        self.assertEqual(find_anagrams(s, p), expected)

    def test_2(self):
        s = "abab"
        p = "ab"
        expected = [0, 1, 2]
        self.assertEqual(find_anagrams(s, p), expected)

    def test_3(self):
        s = "nabanabannaabbaanana"
        p = "banana"
        expected = [0, 3, 5, 6, 7, 13]
        self.assertEqual(find_anagrams(s, p), expected)

if __name__ == "__main__":
    sys.exit(unittest.main())

File: leetcode/find_anagrams.cpp
#include <map>
#include <string>

// number: 438
// similar: 567
// title: Find All Anagrams in a String
// url: https://leetcode.com/problems/find-all-anagrams-in-a-string/
// section: meta
// difficulty: medium
// tags: hash table, string, sliding window, meta, grind 75

// constraints
// 1 <= s.length, p.length <= 3 * 10^4
// s and p consist of lowercase English letters.

// solution: LeetCode solution, sliding window + defaultdict
// complexity:
// run-time: O(n^2)
// space: O(1) (only 26 letters per map)
std::vector<int> findAnagrams(const std::string& s, const std::string& p) {
    std::vector<int> ans;

    if (p.size() > s.size()) {
        return ans;
    }

    // sliding window counter
    std::map<char, int> pmap;
    for (size_t i = 0; i != p.size(); ++i) {
        auto it = pmap.find(p[i]);
        if (it != pmap.end()) {
            pmap[p[i]] += 1;
        } else {
            pmap[p[i]] = 1;
        }
    }

    std::map<char, int> winmap;
    int left = 0;

    for (size_t right = 0; right != s.size(); ++right) {
        auto it = winmap.find(s[right]);
        if (it != winmap.end()) {
            winmap[s[right]] += 1;
        } else {
            winmap[s[right]] = 1;
        }

        // window is smaller than p
        if (right-left+1 < p.size()) {
            continue;
        }

        // check if the current window is a permutation of p
        if (winmap == pmap) {
            ans.push_back(left);
        }

        // remove the leftmost character from the window
        if (winmap[s[left]] > 1) {
            winmap[s[left]] -= 1;
        } else {
            winmap.erase(s[left]);
        }

        // sliding the window
        ++left;
    }

    return ans;
}

// TODO: add unit tests